cli/compose/types/types.go

```diff
@@ -359,46 +497,61 @@ func (u *UlimitsConfig) MarshalYAML() (interface{}, error) {
 	return u, nil
 }
 
+// MarshalJSON makes UlimitsConfig implement json.Marshaller
+func (u *UlimitsConfig) MarshalJSON() ([]byte, error) {
```

Comment:
Why don't you add also the `UnmarshalJSON` ? If read this code correctly, it either produce a simple int `18`, or a json map `{ "soft": 17, "hard": 12}`. I don't think the default JSON unmarshaller will parse these 2 results. It will expect a map with up to 3 keys.

Reply:
For the scope of this PR, we're not really interested in parsing JSON using the `encoding/json` library, as we can use the existing code to parse the JSON file as if it were a YAML file. See the [YAML spec](http://yaml.org/spec/1.2/spec.html#id2759572) on the subject:

> YAML can therefore be viewed as a natural superset of JSON [...]. This is also the case in practice; every JSON file is also a valid YAML file.

Reply:
Doesn't the following work ? (and would be closer to `MarshalYAML`)

```go
if u.Single != 0 {
    return json.Marshal(u.Single)
}
return json.Marshal(u)
```

… as we can make the assumption `Single` is `0`, this won't be marshalled with a `0` value.

Reply:
`json.Marshal(u)` will call `u.MarshalJSON()`, thus recursing infinitely, so no.

Reply:
I don't think I understand why UnmarshalJSON/Yaml was added

Reply:
Then

```go
if u.Single != 0 {
    return json.Marshal(u.Single)
}
return json.Marshal(&UlimitsConfig{Hard: u.Hard, Soft: u.Soft})
```

Reply:
https://play.golang.org/p/X1HpAqXsJHZ 🤷

Reply:
Black pointer magic >_>

😛

Reply:
Oh it actually make sense :angel: (because `MarshalJSON` is not implemented on `UlimitsConfig` :angel:

Reply:
e7f43f1d8f32aa49bff3ec84b06322fd823630e4 👍 

I get why it works, but it's probably going to be obscure to most people IMO especially if they're not familiar with Go